### PR TITLE
asyncio: #10 shutdown_asyncgens

### PR DESCRIPTION
**Pull request by delfick**
Ensure any async generators that are left running past
run_until_complete (essentially only if we get a KeyboardInterrupt that
stops the run_until_complete) then those are finalized properly

Note I don't use loop.shutdown_asyncgens sothat I can make the generator
handle an asyncio.CancelledError rather than a GeneratorExit

### CODE REVIEW

File: tests/test_override_loop.py [hunk 1]
```diff
@@ -133,6 +135,69 @@ async def blah():
...
+        await outside_gen.__anext__()
+
+    original.run_until_complete(outside1())
+    assert info1 == [1, 2]
+
+    with OverrideLoop(new_loop=True) as custom_loop:
+        assert info2 == []
+
+        async def doit():
+            ag = my_generator(info2)
```
[the remark this answers is not in the document]
Comment (nzig) [reply tests/test_override_loop.py:171]: OK, I tried this out myself and it's a a little complicated, but here's an example of a failure case without this PR:
```python
import asyncio
from alt_pytest_asyncio.plugin import OverrideLoop

info2 = []


async def my_generator(info):
    try:
        info.append(1)
        yield
        info.append(2)
        yield
        info.append(3)
    except asyncio.CancelledError:
        info.append("cancelled")
        raise
    finally:
        info.append(("done", __import__("sys").exc_info()[0]))


ag = my_generator(info2)


async def doit():
    await ag.__anext__()
    await ag.__anext__()


with OverrideLoop(new_loop=True) as custom_loop:
    custom_loop.run_until_complete(doit())

assert info2 == [1, 2, "cancelled", ("done", asyncio.CancelledError)]
```

The key here is defining `ag = my_generator(info2)` outside. The reason is that asyncio uses `sys.set_asyncgen_hooks()` to set a finalizer that will `athrow()` any async generator _when they are GC'd_. Because there are no references to `ag` after `doit()` is done, it will probably get garbage collected (but this is not guaranteed), and then asyncio's finalizer will create a task to do `athrow()` on the async generator. This task will then get cancelled by `cancel_all_tasks()`, causing the test to pass without `os.kill()`.

In other words, the test may fail occasionally without `os.kill()`, but this is up to GC. By forcing `ag` to remain alive we can get it to fail every time.

[the remark this answers is not in the document]
Comment (delfick) [reply tests/test_override_loop.py:171]: on looking at that I'm confused why it wouldn't complain that the generator is being created with a different loop to the one you're running. But I suppose that firstiterhook doesn't get called till the first `__anext__` so it could make sense that making the async generator object itself doesn't do anything with the loop.

I'll have a play later. If that removes the os.kill that'd be nice.

[the remark this answers is not in the document]
Comment (delfick) [reply tests/test_override_loop.py]: >  asyncio.run(g.__anext__())
> 1
> asyncio.run(g.__anext__())
> 2

Ever since 2016 for me has been a constant fight with asyncio with because of this project https://github.com/delfick/photons (and specifically the non open source version it's based off)

I'm a big fan of asyncio, but it has some weird edges that cause annoying surprises. The above surprises me because of how angry it gets when you have a future from one loop used by another. So I imagine it'd be fairly simple to make a generator where it complains about the above.

[the remark this answers is not in the document]
Comment (nzig) [reply tests/test_override_loop.py]: Yes, this solves the problem that led to #10. Thanks!

File: tests/test_override_loop.py [hunk 2]
```diff
@@ -133,6 +133,78 @@ async def blah():
...
+            assert set(asyncio.get_event_loop()._asyncgens) == set()
+            await ag2.__anext__()
+            assert set(asyncio.get_event_loop()._asyncgens) == set([ag2])
+            await ag.__anext__()
+            assert set(asyncio.get_event_loop()._asyncgens) == set([ag2, ag])
+            await ag.__anext__()
+            assert info3 == [1]
+
+        custom_loop.run_until_complete(doit())
+        assert list(custom_loop.loop._asyncgens) == [ag]
```
Comment (nzig) [inline tests/test_override_loop.py:187]: This depends on the implementation of the GC, because it assumes `ag2` is collected before this line. That's probably fine given how CPython's GC works right now, but it's not guaranteed to stay the same, so this test might break in a future version of Python (or even current implementations of Python other than CPython).

File: tests/test_override_loop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (delfick) [reply tests/test_override_loop.py]: wicked awesome! I assume I'll make a release next week. (and if #9 isn't ready by then I'll make a release without that)

**Comment by delfick** [comment]
I decided I had time this morning to push this out and you can find it in 0.6.0 :)
